Refactor Connection UI for Multichain

adonesky1 opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/31326?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

```
const EXTENSION_ID = 'pmeejofbihagkmnpoeoghmdmpaonndpl'; // replace this with your local extension ID
const extensionPort = chrome.runtime.connect(EXTENSION_ID)



// EVM only, no accounts requested
// default account: last selected EVM account
// default chains: the selected chain
// existing default EVM chain logic (ONLY APPLIES TO EIP-1193 FLOW since no longer possible to request empty scopes)
extensionPort.postMessage({
    type: 'caip-x',
    data: {
        "jsonrpc": "2.0",
        method: 'wallet_createSession',
        params: {
            optionalScopes: {
                'eip155:1': {
                    methods: [],
                    notifications: []
                },
            },
        },
    }
})


// Solana only, 2 accounts requested
// default accounts: selected accounts that exist/valid. If none, last selected account for that namespace
// default chains: the selected chain
extensionPort.postMessage({
    type: 'caip-x',
    data: {
        "jsonrpc": "2.0",
        method: 'wallet_createSession',
        params: {
            optionalScopes: {
                'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp': {
                    methods: [], 
                    notifications: [],
                    accounts: ['solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp:<solana_account_1>',
                                'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp:<solana_account_2>']
                }
            },
        },
    }
})

// Solana only, no accounts requested
// default accounts: Last selected account for that namespace
extensionPort.postMessage({
    type: 'caip-x',
    data: {
        "jsonrpc": "2.0",
        method: 'wallet_createSession',
        params: {
            optionalScopes: {
                'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp': {
                    methods: [], 
                    notifications: [],
                    accounts: []
                }
            },
        },
    }
})

// EVM and Solana, 2 solana accounts requested
// for provided, selected accounts that exist/valid. At least one account for each namespace (the last selected for that namespace)
extensionPort.postMessage({
    type: 'caip-x',
    data: {
        "jsonrpc": "2.0",
        method: 'wallet_createSession',
        params: {
            optionalScopes: {
                'eip155:1': {
                    methods: [],
                    notifications: []
                },
                'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp': {
                    methods: [],
                    notifications: [],
                    accounts: ['solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp:<solana_account_1>',
                                'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp:<solana_account_2>']
                }
            },
        },
    }
})
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

adonesky1 committed:
Refactor Connection UI for Multichain

adonesky1 committed:
cleanup

adonesky1 committed:
cleaning

adonesky1 committed:
fix

adonesky1 committed:
cleanup

adonesky1 committed:
fix circular dep

adonesky1 committed:
lint

adonesky1 committed:
more progress

adonesky1 committed:
more

jiexi commented on shared/constants/network.ts:
```diff
@@ -818,7 +820,7 @@ export const CHAIN_ID_TO_RPC_URL_MAP = {
 } as const;
 
-export const CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP: Record<Hex, string> = {
+export const CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP: Record<Hex | string, string> = {
```
how awful would it be to keep this constant and create a new constant with a `Record<CaipChainId, string>` type that maps this existing evm `CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP` values and adds the new solona one?

jiexi replied:
i imagine we will want to do this eventually, but it's fine if that time is not now

jiexi replied:
oh boy... I see we aren't the first tack this on :______) . Well then. I feel less guilt since other parties have already taken to bandaging

jiexi commented on shared/modules/selectors/networks.ts:
```diff
@@ -43,6 +44,60 @@ export function getSelectedNetworkClientId(
+/**
+ * Type to extend InternalNetworkState with multichain configurations.
+ */
+type ExtendedInternalNetworkState = InternalNetworkState & {
```
any idea why `NetworkState` doesn't export `multichainNetworkConfigurationsByChainId`?

jiexi replied:
oh... that's because it comes from the MultichainNetworkController.

https://github.com/MetaMask/core/blob/ec6967563ea6254f38ad14e81baa7a46a7b33636/packages/multichain-network-controller/src/types.ts#L84C13-L84C45

jiexi replied:
working on this

jiexi commented on shared/modules/selectors/networks.ts:
```diff
@@ -43,6 +44,60 @@ export function getSelectedNetworkClientId(
+/**
+ * State type that includes both standard and multichain network configurations.
+ */
+export type ConsolidatedNetworkConfigurationsState = {
```
curious what narrowing down the type really buys us. I imagine easier test setup, but not sure benefits past that

jiexi commented on shared/modules/selectors/networks.ts:
```diff
@@ -43,6 +44,60 @@ export function getSelectedNetworkClientId(
+ * @param state - Redux state.
+ * @returns A consolidated object containing all available network configurations.
+ */
+export const getAllNetworkConfigurationsByCaipChainId = createDeepEqualSelector(
```
this begs the question, should the NetworkController also keep evm networkconfigurations in caip2 format in the `multichainNetworkConfigurationsByChainId` in addition to the usual `networkConfigurationsByChainId` so that we can start migrating to this new property in state

jiexi commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.test.tsx:
```diff
@@ -40,10 +41,21 @@ const render = (
   ) as unknown as MergedInternalAccount[];
 
+  const accountsWithCaipAccountId = accounts.map((account) => {
+    const { namespace, reference } = parseCaipChainId(account.scopes[0]);
```
do we need to account for when scopes contain more scopes? also is it necessary to parse the scope into namespace and reference when you could just contact it as is to the address? `${account.scopes[0]}:${account.address}`

jiexi replied:
ehhhhh it's just for fixtures

jiexi commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -38,10 +39,13 @@ import { MetaMetricsContext } from '../../../contexts/metametrics';
-  accounts: MergedInternalAccount[];
-  defaultSelectedAccountAddresses: string[];
+  accounts: {
+    internalAccount: MergedInternalAccount;
+    caipAccountId: CaipAccountId;
```
thoughts on throwing these into a type?

jiexi commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -66,21 +70,28 @@ export const EditAccountsModal: React.FC<EditAccountsModalProps> = ({
-  const handleAccountClick = (address: string) => {
-    if (selectedAccountAddresses.includes(address)) {
+  const handleAccountClick = (caipAccountId: CaipAccountId) => {
+    console.log('caipAccountId', caipAccountId);
```
```suggestion
```

jiexi commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -66,21 +70,28 @@ export const EditAccountsModal: React.FC<EditAccountsModalProps> = ({
+  const handleAccountClick = (caipAccountId: CaipAccountId) => {
+    console.log('caipAccountId', caipAccountId);
+    if (selectedAccountAddresses.includes(caipAccountId)) {
+      console.log('removing caipAccountId', caipAccountId);
```
```suggestion
```

jiexi commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -66,21 +70,28 @@ export const EditAccountsModal: React.FC<EditAccountsModalProps> = ({
       );
     } else {
-      setSelectedAccountAddresses([...selectedAccountAddresses, address]);
+      console.log('adding caipAccountId', caipAccountId);
```
```suggestion
```

adonesky1 committed:
more

adonesky1 committed:
rename selector

jiexi committed:
comment out remaining getPermittedAccountsForCurrentTab errors

jiexi committed:
fix selected account

jiexi committed:
Fix review permissions permitted accounts

adonesky1 committed:
fix

jiexi committed:
Fix edit networks review permissions flow

jiexi committed:
Merge branch 'ad/chain-agnostic-permission-connection-ui-from-top' of github.com:MetaMask/metamask-extension into ad/chain-agnostic-permission-connection-ui-from-top

jiexi committed:
fix add and remove chains

jiexi committed:
Fix add remove permitted accounts permission page

jiexi commented:
TODO: Add support for evaluating supported accounts into the handler. Do the same that we do for EVM accounts, generically for all scope accounts

jiexi committed:
Fix connect page networks

jiexi committed:
Fix lastSelected default account

jiexi commented on ui/components/app/connected-status-indicator/connected-status-indicator.js:
```diff
@@ -31,7 +30,9 @@ export default function ConnectedStatusIndicator({ onClick, disabled }) {
     .includes(WALLET_SNAP_PERMISSION_KEY);
 
-  const permittedAccounts = useSelector(getPermittedAccountsForCurrentTab);
+  // TODO: Fix this
+  // const permittedAccounts = useSelector(getPermittedAccountsForCurrentTab);
+  const permittedAccounts = [];
```
TODO

jiexi commented on ui/index.js:
```diff
@@ -129,8 +128,10 @@ export async function setupInitialStore(
     const { origin } = draftInitialState.activeTab;
-    const permittedAccountsForCurrentTab =
-      getPermittedAccountsForCurrentTab(draftInitialState);
+    // TODO: Fix this
+    // const permittedAccountsForCurrentTab =
+    //   getPermittedAccountsForCurrentTab(draftInitialState);
+    const permittedAccountsForCurrentTab = [];
```
TODO

jiexi commented on ui/selectors/selectors.js:
```diff
@@ -3425,12 +3447,51 @@ function getCaveatFromPermission(caip25Permission = {}) {
+    caip25Caveat.value.optionalScopes,
+  )
+
+  // TODO: May need to filter out wallet scopes here?
```
TODO: verify `wallet` scopes don't get wiped

jiexi replied:
checking this now

jiexi replied:
don't seem to get wiped for some reason. so my assumption is wrong somewhere

jiexi commented on ui/store/actions.ts:
```diff
@@ -1844,8 +1843,10 @@ export function setSelectedAccount(
     const internalAccount = getInternalAccountByAddress(state, address);
-    const permittedAccountsForCurrentTab =
-      getPermittedAccountsForCurrentTab(state);
+    // TODO: fix this
+    // const permittedAccountsForCurrentTab =
+    //   getPermittedAccountsForCurrentTab(state);
+    const permittedAccountsForCurrentTab = [];
```
TODO

jiexi replied:
see https://github.com/MetaMask/metamask-extension/pull/31326#discussion_r2019121259

jiexi committed:
fix connected status indicator

jiexi committed:
note broken toast master

jiexi commented on ui/components/app/toast-master/toast-master.js:
```diff
@@ -110,6 +110,7 @@ export function ToastMaster() {
   return null;
 }
 
+// TODO: Fix this file
```
TODO

jiexi replied:
and elsewhere `addPermittedAccount` or any background-js method is called

jiexi replied:
and removePermittedAccounts 🥲

jiexi replied:
reverted add/removePermittedAccount(s) back to accepting plain hex addresses

jiexi committed:
fix UI index current account is connected

jiexi commented on ui/store/actions.ts:
```diff
@@ -1844,8 +1843,10 @@ export function setSelectedAccount(
+    // TODO: fix this
+    // const permittedAccountsForCurrentTab =
+    //   getPermittedAccountsForCurrentTab(state);
+    const permittedAccountsForCurrentTab = [];
```
the logic in this helper is bugged in main. notice below how `currentTabIsConnectedToPreviousAddress` and `currentTabIsConnectedToNextAddress` effectively are checking the same address. Likely `internalAccount` should be  getting the account for the active tab instead

jiexi committed:
Fix toastmaster

jiexi committed:
Revert back to excepting address instead of CaipAccountId in add/removePermittedAccount(s) background api

jiexi commented on ui/components/multichain/connect-accounts-modal/connect-accounts-modal-list.tsx:
```diff
@@ -28,6 +28,7 @@ import { getURLHost } from '../../../helpers/utils/util';
 import { addPermittedAccounts } from '../../../store/actions';
 import { ConnectAccountsListProps } from './connect-account-modal.types';
 
+// check usage of this component
```
TODO

jiexi replied:
used on the ConnectionsPage. still need to verify deeper

adonesky1 replied:
looking at this

adonesky1 replied:
Seeing it in `routes.component.js` but unsure if this route is ever used now?

Asking @darkwing : https://consensys.slack.com/archives/C06FXU326RL/p1743616755737269

jiexi committed:
remove fix this comment

jiexi committed:
lint

jiexi committed:
Merge branch 'sip-26' into ad/chain-agnostic-permission-connection-ui-from-top

jiexi committed:
Fix ConnectionListItem permitted chain count

jiexi committed:
fix EditAccountsModal test

jiexi committed:
Fix SiteCellTooltip test

jiexi committed:
Merge branch 'sip-26' into ad/chain-agnostic-permission-connection-ui-from-top

jiexi committed:
Fix connect page spec

jiexi committed:
Merge branch 'sip-26' into ad/chain-agnostic-permission-connection-ui-from-top

jiexi committed:
Fix default accounts for wallet:eip155

jiexi committed:
do not show multichain networks without account

jiexi committed:
Fix address case sensitivity checks

jiexi committed:
Fix default network test net check

jiexi committed:
Fix connected network count on review permissions page

jiexi committed:
lint

jiexi committed:
lint

adonesky1 committed:
fix failing test

jiexi committed:
fix background-api test

jiexi committed:
fix edit accounts modal storybook

jiexi committed:
fix wallet in requestedNamespaces

jiexi committed:
temp disable api-specs-mulitchain conf wallet_createSession

jiexi committed:
lint

jiexi committed:
add comment to handler

jiexi committed:
Fix EditAccountsModal storybook actually

jiexi commented on app/scripts/lib/rpc-method-middleware/handlers/wallet-createSession/handler.ts:
```diff
@@ -172,6 +172,7 @@ async function walletCreateSessionHandler(
       sessionProperties: filteredSessionProperties,
     };
 
+    // TODO: we need to apply this same filtering generically to non-evm accounts as well!
```
this is ticketed

jiexi commented on app/scripts/controllers/permissions/background-api.js:
```diff
@@ -196,13 +270,19 @@ export function getPermissionBackgroundApiMethods({
       }
 
-      const existingEthChainIds = getPermittedEthChainIds(caip25Caveat.value);
+      // TODO dry and or move to @metamask/chain-agnostic-permission
+      const requiredScopes = Object.keys(caip25Caveat.value.requiredScopes);
+      const optionalScopes = Object.keys(caip25Caveat.value.optionalScopes);
+
+      const existingChainIds = Array.from(
+        new Set([...requiredScopes, ...optionalScopes]),
+      );
```
let's not block on this. we'll deal with it post merge into the sip-26 branch

jiexi commented on app/scripts/controllers/permissions/background-api.js:
```diff
@@ -147,20 +185,56 @@ export function getPermissionBackgroundApiMethods({
-      const remainingAccounts = existingAccounts.filter(
-        (existingAccount) => existingAccount !== account,
+      const permittedAccounts = new Set();
+      Object.values(caip25Caveat.value.requiredScopes).forEach(
+        ({ accounts }) => {
+          accounts.forEach((account) => {
+            permittedAccounts.add(account);
+          });
+        },
+      );
+      Object.values(caip25Caveat.value.optionalScopes).forEach(
+        ({ accounts }) => {
+          accounts.forEach((account) => {
+            permittedAccounts.add(account);
+          });
+        },
       );
 
+      const existingAccounts = Array.from(permittedAccounts);
```
this logic needs to be DRY'ed into a new getter in core, but again, let's not block on this merging into sip-26 branch

adonesky1 replied:
Should we add a todo comment inline to remind us?

jiexi commented on app/scripts/controllers/permissions/background-api.js:
```diff
@@ -147,20 +185,56 @@ export function getPermissionBackgroundApiMethods({
+            return false;
+          }
+
+          return reference === '0' || reference === existingReference;
```
`reference === '0'` is scuffed... we have the worst of both worlds here having to handle both `eip155:0` and `wallet:eip155`

adonesky1 replied:
alas yes. Any recommendations as to how to proceed for now?

jiexi commented on app/scripts/controllers/permissions/background-api.js:
```diff
@@ -38,23 +39,47 @@ export function getPermissionBackgroundApiMethods({
+    const caipAccountIds = internalAccounts.map((internalAccount) => {
+      const { namespace, reference } = parseCaipChainId(
+        internalAccount.scopes[0],
+      );
```
not iterating over every single scope is okay for now since `setPermittedAccounts` will set the account on each matching namespace anyway, but this doesn't make me happy still

jiexi commented on app/scripts/controllers/permissions/background-api.test.js:
```diff
@@ -359,15 +477,24 @@ describe('permission background API methods', () => {
     });
 
-    it('revokes the entire permission if the removed account is the only eip:155 scoped account', () => {
+    // TODO: review intended behavior here
+    it('revokes the entire permission if the removed account is the only account', () => {
```
awkwardness here with `wallet` and `wallet:eip155`.

jiexi replied:
This may not matter because the UI force you to explicitly disconnect if all accounts are unselected

jiexi commented on app/scripts/controllers/permissions/background-api.test.js:
```diff
@@ -882,26 +1023,21 @@ describe('permission background API methods', () => {
     });
 
-    it('revokes the entire permission if the removed chain is the only eip:155 scope', () => {
+    // TODO: Verify this behavior (wallet vs non-wallet scopes)
+    it('revokes the entire permission if the removed chain is the only scope', () => {
```
awkwardness here with `wallet` and `wallet:eip155` also since they're considered equal important as other chains with the way things are written now.

jiexi replied:
This may not matter because the UI force you to explicitly disconnect if all chains are unselected

jiexi commented on shared/modules/selectors/networks.ts:
```diff
@@ -43,6 +45,81 @@ export function getSelectedNetworkClientId(
+      },
+    );
+
+    Object.entries(multichainNetworkConfigurationsByChainId).forEach(
+      ([caipChainId, networkConfig]) => {
+        const matchesAccount = Object.values(internalAccounts.accounts).some(
+          (account) => {
+            const matchesScope = account.scopes.some((scope) => {
+              return scope === caipChainId;
+            });
+
+            const isSnapEnabled = account.metadata.snap?.enabled;
+
+            return matchesScope && isSnapEnabled;
+          },
+        );
+
+        if (matchesAccount) {
+          caipFormattedEvmNetworkConfigurations[caipChainId] = networkConfig;
+        }
+      },
+    );
```
should probably leave a comment by this filtering logic is needed and why it filters this way

adonesky1 replied:
https://github.com/MetaMask/metamask-extension/pull/31326/files#r2025160850

jiexi commented on test/e2e/run-api-specs-multichain.ts:
```diff
@@ -151,9 +151,11 @@ async function main() {
-          }),
+          // Temporarily disabled as the wallet/wallet:eip155 behavior is broken
+          // but this shouldn't block Solana integration
+          // new MultichainAuthorizationConfirmation({
```
this is ticketed https://github.com/orgs/MetaMask/projects/146/views/6?pane=issue&itemId=104581070&issue=MetaMask%7CMetaMask-planning%7C4569

jiexi commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -111,22 +118,40 @@ export const ReviewPermissions = () => {
+    getAllPermittedChainsForSelectedTab(state, activeTabOrigin),
+  ) as CaipChainId[];
+
+  const connectedChainIds = _connectedChainIds.filter((chainId) => {
+    // plain "wallet" value will break this
+    const { namespace } = parseCaipChainId(chainId);
+    return namespace !== KnownCaipNamespace.Wallet;
+  });
```
fix this 🥲

jiexi commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -145,28 +170,89 @@ export const ReviewPermissions = () => {
 
-  const handleSelectAccountAddresses = (addresses: string[]) => {
-    if (addresses.length === 0) {
+  // should this be at the selector level? :|
+  const connectedAccountAddresses = uniq(
+    _connectedAccountAddresses.map((caipAccountId) => {
+      const {
+        address,
+        chain: { namespace, reference },
+      } = parseCaipAccountId(caipAccountId);
+      if (namespace === KnownCaipNamespace.Eip155) {
+        // this is very hacky, but it works for now
+        return `${namespace}:0:${address}` as CaipAccountId;
+      }
+      return `${namespace}:${reference}:${address}` as CaipAccountId;
+    }),
+  );
```
I think massaging this this way is fine for now. but should we not be doing similar for Solana?..

Here, an EVM address assigned to multiple scopes gets uniqed into `eip155:0:address`, but the non evm addresses assigned to multiple scopes do not get uniqed and one will exist for each scope

adonesky1 replied:
Looking at this

jiexi commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -145,28 +170,89 @@ export const ReviewPermissions = () => {
-    connectedAccountAddresses.forEach((address: string) => {
-      if (!addresses.includes(address)) {
-        dispatch(removePermittedAccount(activeTabOrigin, address));
+    connectedAccountAddresses.forEach((connectedAddress: string) => {
+      const parsedConnectedAddress = parseCaipAccountId(
+        connectedAddress as CaipAccountId,
+      );
+
+      const includesCaipAccountId = caipAccountIds.some((caipAccountId) => {
+        const parsedAddress = parseCaipAccountId(caipAccountId);
+
+        if (
+          parsedConnectedAddress.chain.namespace !==
+            parsedAddress.chain.namespace ||
+          !isEqualCaseInsensitive(
+            parsedConnectedAddress.address,
+            parsedAddress.address,
+          )
+        ) {
+          return false;
+        }
+
+        return (
+          parsedAddress.chain.reference === '0' ||
+          parsedAddress.chain.reference ===
+            parsedConnectedAddress.chain.reference
+        );
+      });
```
this logic is similarly shaped to other logic in this PR. perhaps can be DRYed

jiexi commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -178,6 +264,8 @@ export const ReviewPermissions = () => {
     setShowNetworkToast(false);
   };
 
+  console.log({ connectedAccountAddresses });
+
```
```suggestion
```

jiexi commented on ui/components/multichain/pages/review-permissions-page/site-cell/site-cell-tooltip.test.js:
```diff
@@ -13,97 +13,103 @@ describe('SiteCellTooltip', () => {
-          nameLastUpdatedAt: 1727088231912,
-          importTime: 1727088231225,
-          lastSelected: 1727088231278,
+        internalAccount: {
```
need to undo this {internalAccount, caipAccountId} type

jiexi replied:
i got this

jiexi commented on ui/components/multichain/pages/review-permissions-page/site-cell/site-cell.tsx:
```diff
@@ -59,38 +63,42 @@ export const SiteCell: React.FC<SiteCellProps> = ({
+    selectedChainIds.includes(caipChainId),
   );
 
+  console.log({ allNetworks, selectedNetworks, selectedChainIds });
+
```
```suggestion
```

jiexi commented on ui/components/multichain/pages/review-permissions-page/site-cell/site-cell.tsx:
```diff
@@ -59,38 +63,42 @@ export const SiteCell: React.FC<SiteCellProps> = ({
   const networkMessageNotConnectedState =
     selectedChainIdsLength === 1
-      ? t('requestingForNetwork', [selectedNetworks[0].name])
+      ? t('requestingForNetwork', [selectedNetworks[0]?.name])
```
why are these nullish coalesce needed now?

adonesky1 replied:
seems like they are not. removing

jiexi commented on ui/index.js:
```diff
@@ -129,10 +131,40 @@ export async function setupInitialStore(
+
+    let currentTabIsConnectedToSelectedAddress;
+    if (selectedAccount) {
+      currentTabIsConnectedToSelectedAddress =
+        permittedAccountsForCurrentTab.some((account) => {
+          const parsedPermittedAccount = parseCaipAccountId(account);
+
+          return selectedAccount.scopes.some((scope) => {
+            const { namespace, reference } = parseCaipChainId(scope);
+
+            if (
+              namespace !== parsedPermittedAccount.chain.namespace ||
+              !isEqualCaseInsensitive(
+                selectedAccount.address,
+                parsedPermittedAccount.address,
+              )
+            ) {
+              return false;
+            }
+
+            return (
+              reference === '0' ||
+              reference === parsedPermittedAccount.chain.reference
```
similarly shaped logic else where. Good candidate for core package or helper

jiexi commented on ui/index.js:
```diff
@@ -129,10 +131,40 @@ export async function setupInitialStore(
-      getSelectedInternalAccount(draftInitialState)?.address ?? '';
+      getAllPermittedAccountsForCurrentTab(draftInitialState);
+
+    const selectedAccount = getSelectedInternalAccount(draftInitialState);
```
i guess i'm mainly not understanding why the nullish coasce is needed above

adonesky1 replied:
You mean why the nullish coasce **was** needed above before your changes?

jiexi commented on ui/index.js:
```diff
@@ -129,10 +131,40 @@ export async function setupInitialStore(
 
   if (getEnvironmentType() === ENVIRONMENT_TYPE_POPUP) {
     const { origin } = draftInitialState.activeTab;
+
+    // Not entirely sure if these changes are correct. The original logic kind of confuses me.
```
```suggestion
```

jiexi commented on ui/pages/permissions-connect/connect-page/connect-page.test.tsx:
```diff
@@ -261,27 +275,12 @@ describe('ConnectPage', () => {
-    expect(container).toMatchSnapshot();
-  });
 
-  it('should render a disabled confirm if current account is a non-EVM account', () => {
```
no longer true

jiexi commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -64,13 +66,22 @@ import {
+  getAllRequestedChainIds,
 } from './utils';
 
+// put this here because of some circular dependency issue
+export const caipFormattedTestChains = TEST_CHAINS.map(
+  (chainId) => `eip155:${hexToDecimal(chainId)}`,
+);
```
should park this somewhere else?

adonesky1 replied:
yeah probably. Going to look for a good home for it

adonesky1 replied:
pushing a commit to move it

jiexi commented on ui/components/multichain/network-list-menu/network-list-menu.tsx:
```diff
@@ -333,7 +333,8 @@ export const NetworkListMenu = ({ onClose }: { onClose: () => void }) => {
 
     if (permittedAccountAddresses.length > 0) {
-      dispatch(addPermittedChain(selectedTabOrigin, hexChainId));
+      // TODO: verify this chainId is caip shaped
+      dispatch(addPermittedChain(selectedTabOrigin, chainId));
```
todo

jiexi commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -119,101 +116,170 @@ export const ConnectPage: React.FC<ConnectPageProps> = ({
+    supportedRequestedCaipChainIds.length > 0
+      ? supportedRequestedCaipChainIds
+      : defaultSelectedNetworkList;
+
+  console.log({
+    currentlySelectedNetworkChainId,
+    selectedTestNetwork,
+    defaultSelectedNetworkList,
+    defaultSelectedChainIds,
+  });
```
```suggestion
```

jiexi commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -119,101 +116,170 @@ export const ConnectPage: React.FC<ConnectPageProps> = ({
+  ) as MergedInternalAccount[];
+
+  const allAccountsWithCaipAccountId = allAccounts.map((account) => {
+    // I hope we can reliably use the first scope to determine the namespace
+    const { namespace, reference } = parseCaipChainId(account.scopes[0]);
```
same comment here as in the review permission page

jiexi commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -119,101 +116,170 @@ export const ConnectPage: React.FC<ConnectPageProps> = ({
+      ...Object.keys(requestedCaip25CaveatValue.optionalScopes),
+    ].map((scope) => {
+      const scopeString = scope as InternalScopeString;
+      // This will break if 'wallet' is passed in
```
```suggestion
```

TODO verify there aren't other places we need to fix this

jiexi commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -119,101 +116,170 @@ export const ConnectPage: React.FC<ConnectPageProps> = ({
-  const [selectedChainIds, setSelectedChainIds] = useState(
-    defaultSelectedChainIds,
+  // all requested accounts that are found in the wallet
+  const supportedRequestedAccounts = Array.from(requestedCaipAccountIds).reduce(
```
is `requestedCaipAccountIds` not an array already?

jiexi replied:
checking this

jiexi replied:
made the selector return an array instead of a set and then removed the Array.from 
https://github.com/MetaMask/metamask-extension/pull/31326/commits/c6e54164872e95ce5c772c9c4371759bf274de93

jiexi commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -119,101 +116,170 @@ export const ConnectPage: React.FC<ConnectPageProps> = ({
-      isEvmAccountType(account.type),
-    );
-  }, [accounts]);
+  // TODO use currentAccount as well to determine default accounts
+  // const currentAccount = useSelector(getSelectedInternalAccount);
```
ahhh... hmmm.... we should probably do this still 😬

adonesky1 commented on app/scripts/controllers/permissions/background-api.js:
```diff
@@ -147,20 +185,56 @@ export function getPermissionBackgroundApiMethods({
-
-      const remainingAccounts = existingAccounts.filter(
-        (existingAccount) => existingAccount !== account,
+      const permittedAccounts = new Set();
```
```suggestion
    // TODO dry this into core
      const permittedAccounts = new Set();
```

jiexi commented on ui/pages/permissions-connect/connect-page/utils.ts:
```diff
@@ -70,3 +77,177 @@ export function getCaip25PermissionsResponse(
+ * @param caveatValue - The CAIP-25 caveat value containing scopes.
+ * @returns Array of chain IDs.
+ */
+export function getRequestedChainIds(caveatValue: Caip25CaveatValue): string[] {
```
wait why are we doing this?..

jiexi commented on ui/pages/permissions-connect/connect-page/utils.ts:
```diff
@@ -70,3 +77,177 @@ export function getCaip25PermissionsResponse(
+ * @param supportedRequestedAccounts - The supported requested accounts.
+ * @param allAccounts - All available accounts.
+ */
+export function getDefaultAccounts(
```
this helper needs to also consider the currently selected account, and we need specs

adonesky1 commented on shared/modules/selectors/networks.ts:
```diff
@@ -43,6 +45,81 @@ export function getSelectedNetworkClientId(
+      },
+    );
+
+    Object.entries(multichainNetworkConfigurationsByChainId).forEach(
```
```suggestion
// For now we need to filter out networkConfigurations/scopes without accounts because
// the `endowment:caip25` caveat validator will throw if there are no supported accounts for the given scope
    Object.entries(multichainNetworkConfigurationsByChainId).forEach(
```

adonesky1 commented on ui/components/app/connected-status-indicator/connected-status-indicator.js:
```diff
@@ -14,26 +15,51 @@ import {
-    (account) => account === selectedAddress,
+  const permittedAccounts = useSelector(getAllPermittedAccountsForCurrentTab);
+
+  console.log({ permittedAccounts, selectedAccount });
```
```suggestion
```

adonesky1 commented on ui/components/app/connected-status-indicator/connected-status-indicator.js:
```diff
@@ -14,26 +15,51 @@ import {
+        }
+
+        return (
+          reference === '0' ||
+          reference === parsedPermittedAccount.chain.reference
```
🤮

adonesky1 commented on ui/index.js:
```diff
@@ -129,10 +131,40 @@ export async function setupInitialStore(
+    let currentTabIsConnectedToSelectedAddress;
+    if (selectedAccount) {
+      currentTabIsConnectedToSelectedAddress =
+        permittedAccountsForCurrentTab.some((account) => {
```
```suggestion
TODO: dry and move to @metamask/chain-agnostic-permission package
        permittedAccountsForCurrentTab.some((account) => {
```

adonesky1 committed:
CAIP_FORMATTED_TEST_CHAINS

adonesky1 committed:
Update ui/components/app/connected-status-indicator/connected-status-indicator.js

adonesky1 committed:
CAIP_FORMATTED_EVM_TEST_CHAINS

adonesky1 committed:
CAIP_FORMATTED_EVM_TEST_CHAINS

jiexi committed:
Jl/chain agnostic permission connection UI from top/add caip account id to merged internal account (#31523)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/31523?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

jiexi committed:
Fix bad merge

jiexi committed:
remove console logs

jiexi committed:
revert site cell tooltip test

jiexi committed:
Update shared/modules/selectors/networks.ts

Co-authored-by: Alex Donesky <adonesky@gmail.com>

jiexi committed:
Update ui/index.js

Co-authored-by: Alex Donesky <adonesky@gmail.com>

jiexi committed:
Update app/scripts/controllers/permissions/background-api.js

Co-authored-by: Alex Donesky <adonesky@gmail.com>

jiexi committed:
Update ui/index.js

adonesky1 commented on test/e2e/flask/multichain-api/wallet_createSession.spec.ts:
```diff
@@ -148,7 +148,11 @@ describeBrowserOnly(Browser.CHROME, 'Multichain API', function () {
 
         for (const item of networkListItems) {
-          const network = await item.getText();
+          const networkNameDiv = await item.findElement(
+            By.css('div[data-testid]'),
+          );
+          const network = await networkNameDiv.getAttribute('data-testid');
```
still not sure why this was necessary

adonesky1 commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -145,28 +171,79 @@ export const ReviewPermissions = () => {
-      if (!addresses.includes(address)) {
-        dispatch(removePermittedAccount(activeTabOrigin, address));
+    connectedAccountAddresses.forEach((connectedAddress: string) => {
+      const parsedConnectedAddress = parseCaipAccountId(
```
```suggestion
TOD): DRY this 
      const parsedConnectedAddress = parseCaipAccountId(
```

adonesky1 committed:
addPermittedChain chainId arg is typed correctly

adonesky1 committed:
fix comment

jiexi committed:
Replace ConsolidatedNetworkConfigurationsState

jiexi committed:
Merge branch 'ad/chain-agnostic-permission-connection-ui-from-top' of github.com:MetaMask/metamask-extension into ad/chain-agnostic-permission-connection-ui-from-top

jiexi committed:
fix review permissions page wallet

jiexi committed:
remove comment

adonesky1 committed:
remove comment

jiexi committed:
Fix nonTestNetworks, testNetworks selector types

jiexi committed:
Update ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx

adonesky1 committed:
add comment

adonesky1 committed:
add comment

jiexi committed:
Update ui/pages/permissions-connect/connect-page/connect-page.tsx

adonesky1 commented:
TODO: add functionality to auto add chains when account is manually added without matching chains permitted

adonesky1 commented on ui/components/multichain/connect-accounts-modal/connect-accounts-modal-list.tsx:
```diff
@@ -28,6 +28,7 @@ import { getURLHost } from '../../../helpers/utils/util';
 import { addPermittedAccounts } from '../../../store/actions';
 import { ConnectAccountsListProps } from './connect-account-modal.types';
 
+// check usage of this component
```
```suggestion
```

adonesky1 committed:
Update ui/components/multichain/connect-accounts-modal/connect-accounts-modal-list.tsx

jiexi committed:
make getAllRequestedAccounts return an array instead of set

jiexi committed:
Merge branch 'ad/chain-agnostic-permission-connection-ui-from-top' of github.com:MetaMask/metamask-extension into ad/chain-agnostic-permission-connection-ui-from-top

adonesky1 committed:
remove unnecessary optional chaining

jiexi committed:
Remove getRequestedChainIds

adonesky1 committed:
remove comment

jiexi committed:
Merge branch 'ad/chain-agnostic-permission-connection-ui-from-top' of github.com:MetaMask/metamask-extension into ad/chain-agnostic-permission-connection-ui-from-top

jiexi commented on ui/components/multichain/pages/permissions-page/connection-list-item.js:
```diff
@@ -25,13 +25,13 @@ import {
-  const connectedNetworks = useSelector((state) =>
-    getPermittedChainsForSelectedTab(state, connection.origin),
+  const permittedChains = useSelector((state) =>
+    getAllPermittedChainsForSelectedTab(state, connection.origin),
```
fixing

jiexi committed:
Fix connection list item network count

jiexi committed:
Fix setSelectedAccount action

jiexi committed:
Fix setSelectedAccount
